feat: add option `completefilterfunc` for fuzzy completion

chentoast opened the pull request:
This pr adds a new option `completefilterfunc` (or `cff`), and a new api function for insert completion: `vim.api.nvim_complete(colnr, matches, opts)`.

`completefilterfunc` is a function with the signature

`function(prefix, match) ... return score end`,

where a higher score indicates a better match, and a score of 0 removes the match from the completion menu.

`nvim_complete` is basically `complete()`, but with an optional argument `filterfunc`. `filterfunc` can either be a lua function
or a string referencing a vimscript function, and overrides the `completefilterfunc` option for this completion only. This is so that different plugins can use different fuzzy completion functions.

Here's a demo of it in action:

![filtering](https://user-images.githubusercontent.com/36219739/106374413-41e86600-6338-11eb-9aaf-e86019b785f9.gif)

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -2851,3 +2852,117 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
+/// @param matches: A list of string matches.
+/// @param opts: Dictionary. Possible keys:
+///                 filterfunc: function used for filtering the
+///                     matches. See |nvim_register_filterfunc|.
```
we could add `add = true` for `complete_add()` behavior

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -2851,3 +2852,117 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
+                "expected lua function");
+        goto error;
+      }
+      filterfunc = api_new_luaref(v->data.luaref);
```
the canonical pattern is `filterfunc = v->data.luaref; v->data.luaref = LUA_NOREF`

bfredl commented on src/nvim/edit.c:
```diff
@@ -150,6 +154,22 @@ struct compl_S {
+// global_filterfunc is used for all completion sources,
+// while user_filterfunc is only used to filter matches
+// from vim.api.nvim_complete().
+LuaRef *global_filterfunc = NULL;
```
declare these in `edit.h` instead as `EXTERN LuaRef global_filterfunc INIT(= NO_REF)`, then it is initialized once but available by including `edit.h` anywhere.

chentau replied:
This is helpful, thanks. I'm still very new to C, so I'm not very knowledgeable about things like this.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -2851,3 +2852,117 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
+        goto error;
+      }
+      filterfunc = api_new_luaref(v->data.luaref);
+      user_filterfunc = xmalloc(sizeof(LuaRef));
```
what's the point of the pointer indirection? LuaRef is already a reference (integer index into the registry of values)

chentau replied:
Yes that's a much better way of doing this. I only recently realized that LuaRefs are just integers

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -2851,3 +2852,117 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
+  extern LuaRef *active_filterfunc;
+  LuaRef filterfunc;
+
+  if (matches.type != kObjectTypeArray) {
```
do we intend `matches` to change type later? because otherwise we can just declare `Array` in the arg list

chentau replied:
I just took matches to be an object just for convenience, since we can then just use `object_to_vim` to convert to a vim list.

bfredl replied:
you can do this anyway by using `ARRAY_OBJ` macro. Better to use `Array` in signature to get correct metadata.

chentau commented on src/nvim/api/vim.c:
```diff
@@ -2851,3 +2852,103 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
+///
+/// @param func: function to register for filtering.
+///
+void nvim_register_filterfunc(LuaRef func, Error *err)
```
I tried to take `func` to be an Object instead, but when I do that, I keep getting these errors `cannot convert given type` whenever I try to pass in a function.

bfredl replied:
yes this should be a `LuaRef`. In this case we have to use `api_copy_luaref` for now.

chentau commented on src/nvim/edit.c:
```diff
@@ -3311,7 +3465,7 @@ static int ins_compl_bs(void)
   // Respect the 'backspace' option.
   if ((int)(p - line) - (int)compl_col < 0
       || ((int)(p - line) - (int)compl_col == 0
-          && ctrl_x_mode != CTRL_X_OMNI) || ctrl_x_mode == CTRL_X_EVAL
```
For some reason, pressing backspace cancels `complete()` based completion, instead of simply refreshing the list of candidates. The test in test_popup.vim references changing this behavior as a todo, so I thought I'd just do a drive by change.

bfredl replied:
is this issue still in vim? then it could make sense to break this out to a separate PR, so it is easier to backport it to vim :)

chentau replied:
Makes sense. I've deleted the change

bfredl commented on src/nvim/testdir/test_popup.vim:
```diff
@@ -51,18 +51,6 @@ func Test_popup_complete()
-  " (idea is, C-L inserts the match from the popup menu
-  " but if the menu is closed, it will insert the character <c-l>
-  call feedkeys("aJ\<f5>\<bs>\<c-l>\<esc>", 'tx')
-  call assert_equal(["Januar"], getline(1,2))
```
instead of deleting the test, if we fixed the behavior we should test the new better behavior.

tjdevries commented:
Questions:
- Is `filterfunc` always global? is it possible to be for only a buffer?
- Can `filterfunc` be a string, and therefore a vimL funcref?
- Can `filterfunc` be set / should it be?
- How will this interact with `complete_add` and other related completion items?

I will think some more on this as well. I am excited about this possibility :)

chentau commented:
> Is filterfunc always global? is it possible to be for only a buffer?

filterfunc is global, and I'm actually not so sure whether having it be buffer local is actually useful. To me, the much more useful thing to have would be a way to register different filterfuncs for each completion source, ie. a different filterfunc for built in files completion, another one for tag completion, etc. But that work I think can be delegated to a follow up pr.

> Can filterfunc be a string, and therefore a vimL funcref?

Yeah, this would be good to have. For now, the filterfunc api is lua only, to avoid the extra complexities introduced by having to check whether filterfunc is a lua or viml function, and the different ways for calling them from c code, etc. For now, as a workaround, you can use `vim.fn` to pass a vimscript function.

> Can filterfunc be set / should it be?

Down the line, if we allow filterfunc to be a vimscript function, I don't see why not. However, we should consider what setting the option would look like, especially if down the line we allow for multiple filterfuncs, each registered to a different completion source.

chentau commented:
@bfredl 
There is one edge case that might merit more discussion. Sometimes, when starting completion, `compl_leader` will be set to `NULL`, meaning that no sorting/filtering is done. That means that the matches will be shown in their original order, and the user has to type a key in order to actually activate sorting/filtering. This may be a bit jarring for the user, since typing a key may cause the entire completion list to change suddenly. Do you think that this behavior is fine to keep, or should we change it?

bfredl commented:
@chentau Hmm I think it should always call the sort function, it it is not too hard to implement.

chentau commented on src/nvim/edit.c:
```diff
@@ -2699,120 +2753,220 @@ static void trigger_complete_changed_event(int cur)
+  } else {
+    // qsort is not stable: we don't want to flip two items that have
+    // the same score.
+    return ((compl_score_T *)a)->order - ((compl_score_T *)b)->order;
```
Macos, Linux, and Windows implementations of qsort all have different behaviors for when two items have the same score (Linux uses mergesort under the hood, which is stable, while MacOS and Windows do not). So to make sure that we retain the original order of items for consistency and ease of testing, I have to add this annoying order field to the struct.

bfredl replied:
well qsort() doesn't promise anymore than the specs do (even glibc could change the implementation later). If we want more control over sorting we need to include our own sort algorithm, maybe https://github.com/patperry/timsort

oberblastmeister commented:
Could there also be a way to disable filtering and sorting all together but also still allow the user to type while the completion just stays? This can help with plugins that repeatedly call `vim.api.nvim_complete` and want to filter and sort the matches themselves. If vim and the completion plugin are both sorting this can cause flickering. This doesn't work with the regular `complete` because typing will cause the menu to go away and also cause flickering. Currently this can be done with setting the filter function to always return a number other than 0 but there should be an option to do this.

oberblastmeister commented:
Example using the work around to set filterfunc to
```lua
vim.api.nvim_register_filterfunc(function() return 1 end)
```
will produce this behavior:
![completion_menu](https://user-images.githubusercontent.com/61095988/106954247-85105380-6701-11eb-97e7-d85bf8acd7ee.gif)
Having an option to disable filtering should result in the same behavior.

chentau commented:
@oberblastmeister 

If a plugin wants to take over filtering and sorting from the user, it can just specify the `filterfunc` key of the opts dict like so:

`vim.api.nvim_complete(1, my_completion_entries, {filterfunc = my_custom_filter})`.

`my_custom_filter` will automatically override the currently active filterfunc until completion stops.

oberblastmeister commented:
I understand. I am asking if there is a way to disable it. For example I am working on a plugin that calls complete on insert change and will *internally* sort and then display those using `complete`. I am trying to disable the filterfunc and instead sorting internally. This is why I am trying to disable it as not doing so can cause flickering. This is an example of the plugin:
![auto_completion](https://user-images.githubusercontent.com/61095988/106955675-89d60700-6703-11eb-928b-76b29695df29.gif)
As you can see the completion menu is following the cursor and will update the position when the cursor changes while using filterfunc the menu will not follow the cursor, it will only filter. Not disabling the filterfunc causes flickering (im not sure if you can see it in the gif but it is really noticeable):
![flickery](https://user-images.githubusercontent.com/61095988/106956075-11237a80-6704-11eb-87a8-c2145418d30d.gif)

chentau commented:
It's probably a good idea to allow `nil` in the opts argument, so that `vim.api.nvim_complete(1, matches, {filterfunc=nil})` reverts back to the default behavior of exact prefix matching. As far as I can tell (I may be wrong), the flicker is caused by when there is already a popup visible that then gets reset by calling `nvim_complete` again. Another option is to have a filterfunc that shows all candidates when starting completion, but then shuts down the popupmenu after a letter is typed, leaving it up to the plugin to call `nvim_complete` again:

```lua
filterfunc = function(prefix, match)
  if prefix == "" then -- completion just started
    return 1
  else -- user has typed a letter: treat everything as non-matching so popupmenu doesn't show
    return 0
  end
end
```

Alternatively, you could also look into using `completefunc` with `refresh: always`, in which case vim delegates all control over fetching candidates, filtering, and sorting to the `completefunc`.

In general, though, I'm not the biggest fan of plugins triggering and re-triggering completion, since it causes these awkward flashes where the pum is being opened and closed repeatedly. I think a better idea is to add api functions to manipulate the internal list of current matches, something like: `vim.api.nvim_compl_add_matches, vim.api.nvim_compl_clear_matches`, etc.
Then, a plugin can simply start completion once, and then add and delete matches on `InsertChange` while leaving all display logic up to vim.

oberblastmeister commented:
Allowing filterfunc to be nil is a good idea but what I am proposing is to disable any filterer, even the default exact prefix matching. The result of this should be that the user is typing but nothing is happening on the completion popup like [above](https://github.com/neovim/neovim/pull/13854#issuecomment-773599986). The option can be called `frozen`.

You are partially correct about the flickering. The cause of the flickering is vim trying to filter and then the plugin displaying the results through `nvim_complete`. Allowing to disable filtering would fix this. Calling `nvim_complete` multiple times by itself will not cause flickering. What does cause flickering is when nvim tries to filter but the plugin is also trying to filter. Shutting the popup menu after a key is typed is even worse for flickering because the menu will leave and then the plugin will open the menu again. (this is also the default behavior for the vim function `complete`).

The plugin uses dynamic results every time something is typed and also uses a fuzzy matcher. As I understand filter that this pr proposes can only narrow down results. The fuzzy matcher does not necessarily narrow down matches on each key press, it could add more matches. Then we have to call `nvim_complete` again which is very important to not flicker, which can be alleviated by allowing to disable the filterer. I think that this pr is very cool but I think that it should allow the filterfunc to be disabled to allow more extensibility and different ways of doing completion for plugins.

I understand the concern of a plugin triggering completion multiple times but it doesn't actually cause flickering if the filterer can be delegated to either vim or the plugin, in this case allowing the vim filterer to be disabled.

chentau commented:
@oberblastmeister 
Hm I think I get what you're saying. For me though, I think that the solution (a filterfunc that just returns 1) seems perfectly fine to me, and I don't know if it would actually be useful to add it as an a boolean argument.
For one, I don't know if frozen behavior would be so commonly used (besides your plugin, ofc), as many other plugins wanting similar behavior may instead choose to use `completefunc` with `refresh: always`. And most users would probably not use such a feature for built in completion. Secondly, the fix of setting filterfunc to be a function returning 1 only seems simple and easy enough that it wouldn't save that much effort to instead have something like `frozen = true`.

oberblastmeister commented:
@chentau thank you for your explanation. I guess returning one from a function will work. I am wondering about performance though because vim should still be trying to sort the list even though they are all one which seems unnecessary. Also I don't think just this plugin will make use of that feature because there are a plethora of plugins that auto-trigger completion like `completion-nvim`, `coc.nvim`, `kok.nvim`, `compe-nvim` that all might benefit.

tjdevries commented:
> `vim.api.nvim_complete(1, my_completion_entries, {filterfunc = my_custom_filter})`.

I think doing `filterfunc = false` would be a good way to signal that you don't want to do anything internally. (note: `filterfunc = nil` doesn't do anything in Lua).

I think there are some other APIs @oberblastmeister is talking about that could be very useful for us on rofl.nvim ( which is why I assume you're here @oberblastmeister  :laughing: ) that hopefully can make async filling of the completion list possible and easier for external plugins.

I'll think some more about what else is happening here. :)

oberblastmeister commented:
@tjdevries haha yes this was for the beloved rofl.nvim :smile: `filter_func = false` would be nice because even though you set the filterfunc to `function() return 1 end` it might be slower, telling the api to stop filtering might be good.

chentau commented:
@tjdevries @bfredl  This is ready for review again. I'll squash and rebase after this looks ready to merge, and we've decided on how to address the raised comments.

oberblastmeister commented:
When I call complete in quick succession neovim is crashing
`free(): invalid pointer zsh: abort (core dumped)`

This is the output of `coredumpctl -1 gdb`
```
           PID: 244187 (nvim)
           UID: 1000 (brian)
           GID: 1000 (brian)
        Signal: 6 (ABRT)
     Timestamp: Fri 2021-02-05 21:44:05 EST (1min 20s ago)
  Command Line: ./build/bin/nvim /home/brian/projects/syn/syn.json
    Executable: /home/brian/projects/chentau_neovim/build/bin/nvim
 Control Group: /user.slice/user-1000.slice/session-2.scope
          Unit: session-2.scope
         Slice: user-1000.slice
       Session: 2
     Owner UID: 1000 (brian)
       Boot ID: 40b88ab5904e4b1185dc0cac870cf0a6
    Machine ID: e7354646947746d99a794a0413935caf
      Hostname: oberblastmeister
       Storage: /var/lib/systemd/coredump/core.nvim.1000.40b88ab5904e4b1185dc0cac870cf0a6.244187.1612579445000000.zst
       Message: Process 244187 (nvim) of user 1000 dumped core.

                Stack trace of thread 244187:
                #0  0x00007f54afa84615 raise (libc.so.6 + 0x3d615)
                #1  0x00007f54afa6d862 abort (libc.so.6 + 0x26862)
                #2  0x00007f54afac65e8 __libc_message (libc.so.6 + 0x7f5e8)
                #3  0x00007f54aface27a malloc_printerr (libc.so.6 + 0x8727a)
                #4  0x00007f54afacf64c _int_free (libc.so.6 + 0x8864c)
                #5  0x00005599c3661871 n/a (/home/brian/projects/chentau_neovim/build/bin/nvim + 0x329871)
                #6  0x0000000000000008 n/a (n/a + 0x0)
```

tjdevries commented:
Let us know if you have any questions / need help with something!

chentau commented:
@tjdevries  I think this PR is mostly ready, save for maybe even more tests; I'm just waiting for another round of review from @bfredl , although 0.5 things probably take priority over this. I haven't noticed any crashes.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -2937,3 +2938,106 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts, Erro
+///
+/// @param func: function to register for filtering.
+///
+void nvim_register_filterfunc(LuaRef func, Error *err)
```
- this is just a "set" not "register", so `nvim_set_complete_filter_fn`
- instead of this special-case API function, can we add a 'completefilterfunc' option (analogous to the existing 'completefunc' option)? 
    - it's worth the trouble, to avoid another quasi-options interface.

chentau replied:
Making it an option does sound better, but I'm not sure how this would interact with the local 'filterfunc' option specified in `vim.api.nvim_complete`, since one would be a string referencing a viml function (or a lua function via `v:lua`), and the other would be a LuaRef. If possible, I'd like to avoid passing unions of viml funcref and luarefs around. Do you have any suggestions regarding this?

justinmk replied:
> since one would be a string referencing a viml function (or a lua function via `v:lua`), and the other would be a LuaRef.

no, the idea is that options could support Lua functions.

chentau replied:
I've made 'completefilterfunc' an option now.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -2937,3 +2938,106 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts, Erro
+///                     matches. See |nvim_register_filterfunc|.
+///                 exact: boolean value. If true, disables fuzzy
+///                     finding and reverts to default exact prefix
+///                     matching for this completion.
```
missing `@return` doc

justinmk commented on src/nvim/edit.c:
```diff
@@ -2738,124 +2792,231 @@ static void trigger_complete_changed_event(int cur)
+
+// sort the linked list of completion items based on the scores
+// in the scores array.
+static void sort_completions(int num_items, compl_score_T *scores)
```
use common prefixes. `ins_compl_` looks like what other related functions use.
```suggestion
static void ins_compl_sort(int num_items, compl_score_T *scores)
```

justinmk commented on src/nvim/edit.h:
```diff
@@ -12,6 +13,14 @@
+// from vim.api.nvim_complete().
+EXTERN LuaRef global_filterfunc INIT(= LUA_NOREF);
+EXTERN LuaRef user_filterfunc INIT(= LUA_NOREF);
+EXTERN LuaRef active_filterfunc INIT(= LUA_NOREF);
```
would be nice if these are grouped a `compl_filter` struct. Then all references look like `compl_filter.user_func`, etc.

justinmk commented:
see also:

- https://github.com/neovim/neovim/pull/8184
- https://github.com/neovim/neovim/pull/12995

chentau commented:
@justinmk Was this more along the lines of what you were thinking?

chentau commented:
@bfredl @justinmk Sorry for the multiple pings, but do either of you have time to take a look at this PR and see what it still needs?

justinmk commented on runtime/doc/options.txt:
```diff
@@ -1392,6 +1392,22 @@ A jump table for the options with a short description can be found at |Q_op|.
+
+		function(prefix, entry),
+
+	where 'entry' is a completion candidate, 'prefix' is the text
```
single-quotes in vim :help docs have a special meaning, they always refer to editor options.
```suggestion
	where "entry" is a completion candidate, "prefix" is the text
```

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -2882,3 +2883,85 @@ Array nvim_get_mark(String name, Error *err)
 }
 
+///
+/// Set matches and start insert completion: see |complete()|.
```
For reference, what are the reasons for adding this API function instead of augmenting the existing `complete()`?

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -2882,3 +2883,85 @@ Array nvim_get_mark(String name, Error *err)
+
+  if (HAS_KEY(opts->exact) && HAS_KEY(opts->filterfunc)) {
+    api_set_error(err, kErrorTypeValidation,
+                  "cannot set both 'filterfunc' and 'exact'");
```
is "exact" really needed, or is it trivial to mimic it with a function?

justinmk commented on src/nvim/edit.c:
```diff
@@ -3791,7 +3984,7 @@ static bool ins_compl_prep(int c)
       if (c == Ctrl_E) {
         ins_compl_delete();
-        if (compl_leader != NULL) {
+        if (compl_leader != NULL && STRNCMP(compl_leader, "", 1) != 0) {
```
usual pattern in the codebase is to compare `compl_leader[0] == NUL`

justinmk reviewed (commented):
This looks well-considered and I don't have any big hesitation on it. Nicely done!

Is there an accidental change to `src/nvim/option.c` ?


cc @mrbiggfoot is this useful to you?

chentoast committed:
feat: fuzzy filtering via 'completefilterfunc'

Introduces a new option 'completefilterfunc', to filter insert
completion matches. completefilterfunc must be a function with
signature

  function(prefix, match)

and return a score in the interval [0, 999999). Higher scoring
matches are placed higher up on the completion popup menu, and
a score of zero removes the match from being shown.
'completefilterfunc' is called when starting insert mode completion,
and whenever the user types a character to narrow down the
completion list, allowing for native fuzzy filtering.

This PR also introduces a new api function 'nvim_complete',
which allows for overriding the 'completefilterfunc' option
with another filterfunc for this completion only.

chentau commented:
@justinmk  Do you mind reviewing again? I ended up removing the `exact` keyword, partly to simplify the code and also partly so the "callable container" struct can be more generalizable. Overall, I think this PR is getting pretty close to being ready to merge.

zeertzjq commented:
Maybe this also needs to be documented in `vim_diff.txt`.

fsouza commented:
@chentau do you have plans to pick this back up? I would really like to see this through and I'm considering picking it up to try and get it merged. Any thoughts?

chentau commented:
The PR as it currently stands is almost ready for merge. However, my current goal is to rework this PR into something that fits in with what I proposed in #16488, which is more general and is preferable (in my opinion) to adding a new option - so I think I wouldn't recommend just rebasing and trying to push this version through. If you're willing to start work on the issue I linked, I'd be happy to help out and work together, but otherwise, you might have to wait until the summer, which is when I'll have more time.

fsouza commented:
> The PR as it currently stands is almost ready for merge. However, my current goal is to rework this PR into something that fits in with what I proposed in #16488, which is more general and is preferable (in my opinion) to adding a new option - so I think I wouldn't recommend just rebasing and trying to push this version through. If you're willing to start work on the issue I linked, I'd be happy to help out and work together, but otherwise, you might have to wait until the summer, which is when I'll have more time.

Oh I had completely forgotten about that issue. I think that's more work that I'd be able to put on and the summer is just around the corner, so I can definitely wait. I'll keep an eye on that issue and help with whatever I can. Thank you!

justinmk commented:
matchfuzzy() is somewhat related https://github.com/vim/vim/pull/6932
